server.go

```diff
@@ -2602,6 +2389,75 @@ func newServer(listenAddrs []string, db database.DB, chainParams *chaincfg.Param
 	}
 	s.cpuMiner = newCPUMiner(&policy, &s)
 
+	// Only setup a function to return new addresses to connect to when
+	// not running in connect-only mode.
+	var newAddressFunc connmgr.AddressFunc
+	if len(cfg.ConnectPeers) == 0 {
+		newAddressFunc = func() string {
+			for tries := 0; tries < 100; tries++ {
+				addr := s.addrManager.GetAddress("any")
+				if addr == nil {
+					break
+				}
+
+				// Address will not be invalid, local or unroutable
+				// because addrmanager rejects those on addition.
+				// Just check that we don't already have an address
+				// in the same group so that we are not connecting
+				// to the same network segment at the expense of
+				// others.
+				key := addrmgr.GroupKey(addr.NetAddress())
+				if s.OutboundGroupCount(key) != 0 {
+					continue
+				}
+
+				// only allow recent nodes (10mins) after we failed 30
+				// times
+				if tries < 30 && time.Now().Sub(addr.LastAttempt()) < 10*time.Minute {
+					continue
+				}
+
+				// allow nondefault ports after 50 failed tries.
+				if fmt.Sprintf("%d", addr.NetAddress().Port) !=
+					activeNetParams.DefaultPort && tries < 50 {
+					continue
+				}
+				return addrmgr.NetAddressKey(addr.NetAddress())
+			}
+			return ""
+		}
+	}
+
+	// Create a connection manager.
+	cmgr, err := connmgr.New(&connmgr.Config{
+		RetryDuration: connectionRetryInterval,
+		MaxOutbound:   defaultMaxOutbound,
+		Dial:          btcdDial,
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			sp := s.newOutboundPeer(c.Addr, c.Permanent)
+			if sp != nil {
+				sp.Connect(conn)
```

Comment:
Can we rename `peer.Connect` to something like `peer.AssociateConnection` in a separate PR?  Given there is a `connmgr.Connect` which actually establishes a connection, it seems rather confusing to have the same name on a peer when it doesn't actually connect, rather it just associates an existing connection with it.